### PR TITLE
Support Mocking in DataServiceContext, DataServiceQuery, DataServiceQueryOfT and DataServiceQuerySingleOfT classes

### PR DESCRIPTION
**Pull request by ElizabethOkerio**
### Issues

*This pull request fixes issue #1250.*

### Description

The DataServiceContext, DataServiceQuery, DataServiceQueryOfT and DataServiceQuerySingleOfT classes cannot be mocked as they are. In this PR, I have made these classes' public methods and properties virtual to support mocking. 

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.OData.Client/Build.NetStandard/project.json [hunk 1]
```diff
@@ -3,6 +3,7 @@
   "dependencies": {
     "NETStandard.Library": "1.6.0",
     "System.ComponentModel": "4.3.0",
+    "System.ComponentModel.Annotations": "4.3.0",
```
Comment (g2mula) [inline src/Microsoft.OData.Client/Build.NetStandard/project.json]: Probably not meant for this PR..

File: src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs [hunk 2]
```diff
@@ -478,7 +478,7 @@ internal static PropertyInfo[] GetKeyPropertiesOnType(Type type, out bool hasPro
                 }
             }
 
-            if (newKeyKind == KeyKind.AttributedKey && keyProperties.Count != dataServiceKeyAttribute.KeyNames.Count)
+            if (newKeyKind == KeyKind.AttributedKey && keyProperties.Count != dataServiceKeyAttribute?.KeyNames.Count)
```
Comment (g2mula) [inline src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs]: Probably not meant for this PR..

File: src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs [hunk 3]
```diff
@@ -704,6 +704,12 @@ private static KeyKind IsKeyProperty(PropertyInfo propertyInfo, KeyAttribute dat
             {
                 keyKind = KeyKind.AttributedKey;
             }
+#if !PORTABLELIB
+            else if (propertyInfo.GetCustomAttributes().OfType<System.ComponentModel.DataAnnotations.KeyAttribute>().Any())
+            {
+                keyKind = KeyKind.AttributedKey;              
+            }
+#endif
```
Comment (g2mula) [inline src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs]: Probably not meant for this PR..

File: src/Microsoft.OData.Client/Microsoft.OData.Client.NetStandard.VS2017.csproj [hunk 4]
```diff
@@ -60,6 +60,7 @@
 
   <ItemGroup>
     <PackageReference Include="System.ComponentModel" Version="4.3.0" />
+    <PackageReference Include="System.ComponentModel.Annotations" Version="4.3.0" />
```
Comment (g2mula) [inline src/Microsoft.OData.Client/Microsoft.OData.Client.NetStandard.VS2017.csproj]: Probably not meant for this PR..

File: test/FunctionalTests/Microsoft.OData.Client.Tests/DataServiceContextMockTests.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
+﻿//---------------------------------------------------------------------
```
Comment (g2mula) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/DataServiceContextMockTests.cs]: Probably not meant for this PR..

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Metadata/ClientTypeUtilTests.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
+﻿//---------------------------------------------------------------------
```
Comment (g2mula) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/Metadata/ClientTypeUtilTests.cs]: Probably not meant for this PR..

File: test/FunctionalTests/Microsoft.OData.Client.Tests/packages.config [hunk 7]
```diff
@@ -1,5 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="Castle.Core" version="4.4.0" targetFramework="net45" />
+  <package id="Moq" version="4.13.1" targetFramework="net45" />
+  <package id="System.Runtime.CompilerServices.Unsafe" version="4.5.0" targetFramework="net45" />
+  <package id="System.Threading.Tasks.Extensions" version="4.5.1" targetFramework="net45" />
```
Comment (g2mula) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/packages.config]: Probably not meant for this PR..

Comment (ElizabethOkerio) [reply test/FunctionalTests/Microsoft.OData.Client.Tests/packages.config]: this is meant for this PR

**Pull request closed without merging**